app/channel.go

```diff
@@ -722,38 +722,107 @@ func (a *App) GetTeamSchemeRolesForChannel(teamId string) (string, string, strin
 	if err != nil {
 		return nil, err
 	}
 
-	inhertedGuestRole, err := a.GetRoleByName(inhertedGuestRoleName)
+	inheritedGuestRole, err := a.GetRoleByName(inheritedGuestRoleName)
 	if err != nil {
 		return nil, err
 	}
-	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole)
-	inheritedGuestPermissions := GetChannelModeratedPermissions(inhertedGuestRole)
+
+	inheritedMemberPermissions := GetChannelModeratedPermissions(inheritedMemberRole.Permissions)
+	inheritedGuestPermissions := GetChannelModeratedPermissions(inheritedGuestRole.Permissions)
+
+	for _, moderationPatch := range channelModerationsPatch {
+		if moderationPatch.Roles.Members != nil && *moderationPatch.Roles.Members && !inheritedMemberPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+		if moderationPatch.Roles.Guests != nil && *moderationPatch.Roles.Guests && !inheritedGuestPermissions[*moderationPatch.Name] {
+			return nil, &model.AppError{Message: "Cannot add a permission that is restricted by the team or system permission scheme"}
+		}
+	}
+
+	// Channel has no scheme so create one
+	if channel.SchemeId == nil || len(*channel.SchemeId) == 0 {
+		scheme := &model.Scheme{
+			Name:        model.NewId(),
+			DisplayName: model.NewId(),
+			Scope:       model.SCHEME_SCOPE_CHANNEL,
+		}
+		if scheme, err = a.CreateScheme(scheme); err != nil {
+			return nil, err
+		}
+		channel.SchemeId = &scheme.Id
+		a.UpdateChannelScheme(channel)
+	}
+
+	guestRoleName, memberRoleName, _, _ := a.GetSchemeRolesForChannel(channel.Id)
+	memberRole, err := a.GetRoleByName(memberRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	guestRole, err := a.GetRoleByName(guestRoleName)
+	if err != nil {
+		return nil, err
+	}
+
+	memberRolePatch := memberRole.RolePatchFromChannelModerationsPatch(channelModerationsPatch, "members")
+	guestRolePatch := guestRole.RolePatchFromChannelModerationsPatch(channelModerationsPatch, "guests")
+
+	memberRolePermissionsUnmodified := len(model.ChannelModeratedPermissionsChangedByPatch(inheritedMemberRole, memberRolePatch)) == 0
+	guestRolePermissionsUnmodified := len(model.ChannelModeratedPermissionsChangedByPatch(inheritedGuestRole, guestRolePatch)) == 0
+	if memberRolePermissionsUnmodified && guestRolePermissionsUnmodified {
+		if _, err := a.DeleteScheme(*channel.SchemeId); err != nil {
+			return nil, err
+		}
+		channel.SchemeId = nil
+		a.UpdateChannelScheme(channel)
+		memberRole = inheritedMemberRole
+		guestRole = inheritedGuestRole
+	} else {
+		memberRole.Patch(memberRolePatch)
+		guestRole.Patch(guestRolePatch)
+	}
+
+	return buildChannelModerations(memberRole, guestRole, inheritedMemberRole, inheritedGuestRole), nil
+}
+
+func buildChannelModerations(memberRole *model.Role, guestRole *model.Role, inheritedMemberRole *model.Role, inheritedGuestRole *model.Role) []*model.ChannelModeration {
```

Comment:
used by both get channel moderations and patch moderations 
`memberRole` could be the same as `inheritedMemberRole` in the case that there is no channel moderation scheme

Reply:
I like this "inherited" role terminology, I may adopt it in https://github.com/mattermost/mattermost-server/pull/13813.